### ISSUE TITLE
Mac build on Travis CI is broken

### ISSUE THREAD

**Issue by ChristianBeer** [BOINC/boinc#2660]
I'm currently rearranging the Travis configuration and this invalidates the current build cache (see #2659). This means that dependencies get rebuilt while no other code changes occurred. The Mac build failes when linking `boinc`with this error:
```
Compiling mac_address.cpp
Compiling coproc_sched.cpp
Linking boinc

ld: symbol(s) not found for architecture x86_64

clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
I already verified that the dependencies have the correct architectures:
```
input file /Users/travis/build/BOINC/boinc/3rdParty/buildCache/mac/lib/libssl.a is not a fat file
Non-fat file: /Users/travis/build/BOINC/boinc/3rdParty/buildCache/mac/lib/libssl.a is architecture: x86_64

input file /Users/travis/build/BOINC/boinc/3rdParty/buildCache/mac/lib/libcares.a is not a fat file
Non-fat file: /Users/travis/build/BOINC/boinc/3rdParty/buildCache/mac/lib/libcares.a is architecture: x86_64

input file /Users/travis/build/BOINC/boinc/3rdParty/buildCache/mac/lib/libcurl.a is not a fat file
Non-fat file: /Users/travis/build/BOINC/boinc/3rdParty/buildCache/mac/lib/libcurl.a is architecture: x86_64

input file /Users/travis/build/BOINC/boinc/3rdParty/buildCache/mac/lib/libwx_osx_cocoa_static.a is not a fat file
Non-fat file: /Users/travis/build/BOINC/boinc/3rdParty/buildCache/mac/lib/libwx_osx_cocoa_static.a is architecture: x86_64

input file /Users/travis/build/BOINC/boinc/3rdParty/buildCache/mac/lib/libsqlite3.a is not a fat file
Non-fat file: /Users/travis/build/BOINC/boinc/3rdParty/buildCache/mac/lib/libsqlite3.a is architecture: x86_64

Architectures in the fat file: /Users/travis/build/BOINC/boinc/3rdParty/buildCache/mac/lib/libfreetype.a are: i386 x86_64 

Architectures in the fat file: /Users/travis/build/BOINC/boinc/3rdParty/buildCache/mac/lib/libftgl.a are: i386 x86_64
```
Since the only thing that changed is the cache and that seems to be in order I don't have a clue where to look at next. The xcode output from Travis CI is filtered through xcpretty because otherwise we hit the size limit of the logfile.

To recreate this on a Mac one simply has to clone the boinc repository into a fresh directory and execute:
```
./3rdParty/buildMacDependencies.sh -q && ./mac_build/buildMacBOINC-CI.sh --no_shared_headers
```
This should do the same as in the Travis CI VM. If you don't have xcpretty installed you'll see the whole xcode output.

I'm specifically mentioning @CharlieFenton and @armstrdj but every @BOINC/committer and @BOINC/contributors  is also welcomed. This issue holds up the integration and unit testing code I plan to add to BOINC.

**Comment by CharlieFenton** [comment]
A fat file has multiple architectures, usually 32-bit (i686) and 64 bit (x86_64), though in the past it sometimes included PowerPC (ppc). It appears that for some reason Travis CI is trying to build the BOINC Manager, Client and default / overview screensaver (_boincscr_) as fat files. I just now synced with master and confirmed that mac_build/boinc.xcodeproj/project.pbxproj builds only the libraries as fat files (i686 and x86_64) and they do not link to any of the libraries listed in the errors. 

It appears to be building the Client as mac_address.cpp and  coproc_sched.cpp are used only in the Client, and _Linking boinc_ confirms that. There are several odd things here. It says `symbol(s) not found for architecture x86_64` yet it complains that the libraries are only that architecture. Also, some of the libraries listed are not used by the client. _libwx_osx_cocoa_static.a_ (wxWidgets) and _libsqlite3.a_ are linked only by the Manager.  _libfreetype.a_ and _libftgl.a_ are linked only by _boincscr_ (Xcode target _ss_app_.)

I won't have time today to look into this further.

**Comment by CharlieFenton** [comment]
> mac_build/boinc.xcodeproj/project.pbxproj builds only the libraries as fat files (i686 and x86_64) and they do not link to any of the libraries listed in the errors.

To clarify:
mac_build/boinc.xcodeproj/project.pbxproj builds only the BOINC libraries as fat files (i686 and x86_64) and they do not link to any of the dependent  libraries listed in the errors.

**Comment by ChristianBeer** [comment]
Yes, the output is from the `lipo` command I added to the script because I wanted to see if a library has a missing architecture. The build in master is still working and the only difference to my failed build in #2659 is the content of `3rdParty/buildCache/mac`. The boinc libraries are rebuilt every time and are not cached. Unfortunately I can't enable more logging on Travis CI because of the logfile limit there. But I could add some commands to `buildMacBOINC-CI.sh` to investigate the state of the build after it failed. But I don't know what to look for.

**Comment by AenBleidd** [comment]
@ChristianBeer, you can make output more verbose and redirect it to file which you can then upload to https://transfer.sh/ to make further investigation.

**Comment by ChristianBeer** [comment]
Thanks for this link, it's exactly what I needed. The build is running right now.

**Comment by ChristianBeer** [comment]
That helped. Here is the list of undefined symbols from https://transfer.sh/SzSh/boinc_xcodebuild.log
```
Ld build/Deployment/boinc normal x86_64
    cd /Users/travis/build/BOINC/boinc/mac_build
    export MACOSX_DEPLOYMENT_TARGET=10.6
    /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -arch x86_64 -isysroot /Applications/Xcode-9.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -L/Users/travis/build/BOINC/boinc/mac_build/build/Deployment -L/Users/travis/build/BOINC/boinc/3rdParty/buildCache/mac/lib -F/Users/travis/build/BOINC/boinc/mac_build/build/Deployment -filelist /Users/travis/build/BOINC/boinc/mac_build/build/boinc.build/Deployment/BOINC_Client.build/Objects-normal/x86_64/boinc.LinkFileList -mmacosx-version-min=10.6 -dead_strip -Xlinker -object_path_lto -Xlinker /Users/travis/build/BOINC/boinc/mac_build/build/boinc.build/Deployment/BOINC_Client.build/Objects-normal/x86_64/boinc_lto.o -L. -lcurl -lcares -lcrypto -lssl -lldap -lz -framework IOKit -framework Carbon -framework Security -framework OpenGL -Xlinker -dependency_info -Xlinker /Users/travis/build/BOINC/boinc/mac_build/build/boinc.build/Deployment/BOINC_Client.build/Objects-normal/x86_64/boinc_dependency_info.dat -o /Users/travis/build/BOINC/boinc/mac_build/build/Deployment/boinc
clang: warning: libstdc++ is deprecated; move to libc++ with a minimum deployment target of OS X 10.9 [-Wdeprecated]
Undefined symbols for architecture x86_64:
  "_idn2_lookup_ul", referenced from:
      _fix_hostname in libcurl.a(libcurl_la-url.o)
  "_idn2_strerror", referenced from:
      _fix_hostname in libcurl.a(libcurl_la-url.o)
  "_idn2_free", referenced from:
      _Curl_disconnect in libcurl.a(libcurl_la-url.o)
      _create_conn in libcurl.a(libcurl_la-url.o)
  "_idn2_check_version", referenced from:
      _curl_version in libcurl.a(libcurl_la-version.o)
      _curl_version_info in libcurl.a(libcurl_la-version.o)
      _fix_hostname in libcurl.a(libcurl_la-url.o)
ld: symbol(s) not found for architecture x86_64
```

**Comment by CharlieFenton** [comment]
This issue does not occur on the Mac I use to build the Mac releases. The difference in the Travis-CI build from that used to build the release Mac clients is that the Travis-CI build environment has  _x86_64-pkg-config_ and _libidn2_, neither of which are required to build the Mac client. I just rebuilt curl-7.58.0 on this Mac, and here is the output corresponding to the snippet above:
`checking for x86_64-pkg-config... no
checking for pkg-config... /usr/local/bin/pkg-config
checking for zlib options with pkg-config... no
`

Without _x86_64-pkg-config_, curl's configure script apparently does not look for _-lidn2_ or _idn2.h_. On my Mac, there is nothing in /usr/local with either _Cellar_ or _libidn2_, so _x86_64-pkg-config_ would not have found it anyway.

I am extremely careful to avoid installing any third party software such as _libidn2_ on my development system. I never `make install` any libraries or other software, lest it make BOINC work differently than it would for users with a "virgin" Mac system. I even keep BOINC's dependent libraries such as _libcurl_ in the same development directory with the BOINC source tree, and only as static libraries built inside their own source trees, and I keep none of this on the system (boot) volume.

For the Travis-CI Mac builds to be meaningful, it must match the build environment used to build the BOINC releases for the Mac. This implies that you will need to remove from the Travis-CI environment all such extraneous software as _libidn2_ which are not part of the standard BOINC Mac build environment described in _mac_build/HowToBuildBOINC_XCode.rtf_.

Please do NOT under any circumstances add `-lidn2` to the BOINC project. It will make the builds fail for me and anyone else who wishes to build the Mac client for themselves and who has not installed the superfluous _Cellar/libidn2_. According to https://www.gnu.org/software/libidn/, 
> Libidn's purpose is to encode and decode internationalized domain names.

**Comment by CharlieFenton** [comment]
Looking at your commits, I see that you are using _brew_, which is another third-party utility. I very strongly urge you to avoid installing or using **any** development tools in the Travis-CI builds that are not provided as standard with either OS X itself or Xcode. On MSWindows, I would say the same about tools not provided as standard with the Windows OS or Visual Studio.

**Comment by ChristianBeer** [comment]
While I was at the gym I thought about this a bit more and now have a better explanation to what is going on.
The behavior of curl is that it tries to detect the presence of libidn2 and use it if found. The Travis OSX VM that was used to built curl probably didn't have libidn2 installed. But since then the VM was updated and libidn2 is automatically provided which makes curl pick it up now that I triggered a rebuild. I'm going to disable libidn2 in the curl build since we didn't have support for this in the past.

But I would also like to have IDN support in BOINC so I'm going to create an issue for this. What we need to do is build a static version of libidn2 that our static build of curl can use. Since we don't do a static build on linux currently it's a bit easier there.

As to the Travis OSX build environment. We are using a multi-purpose VM that already contains a lot of general usage libraries and tools. That means this kind of "well-meant" issues occur from time to time and we just need to find out how to circumvent them. The solution to this was starring in my face earlier when I read a comment on a curl mailing list about libidn2 but I didn't realize it until later.
